src/java.base/share/classes/sun/security/pkcs12/MacData.java

```diff
@@ -72,28 +83,65 @@ class MacData {
         AlgorithmId digestAlgorithmId = AlgorithmId.parse(digestInfo[0]);
         this.digestAlgorithmName = digestAlgorithmId.getName();
         this.digestAlgorithmParams = digestAlgorithmId.getParameters();
+
         // Get the digest.
         this.digest = digestInfo[1].getOctetString();
 
-        // Get the salt.
-        this.macSalt = macData[1].getOctetString();
+        if (digestInfo[0].tag != DerValue.tag_Sequence) {
+            throw new IOException("algid parse error, not a sequence");
+        }
+        if (digestAlgorithmName.equals("PBMAC1")) {
+            PBEParameterSpec pbeSpec;
+
+            try {
+                pbeSpec =
+                        digestAlgorithmParams.getParameterSpec(
+                        PBEParameterSpec.class);
+            } catch (InvalidParameterSpecException ipse) {
+                throw new IOException(
+                        "Invalid PBE algorithm parameters");
+            }
+            iterations = pbeSpec.getIterationCount();
+            macSalt = pbeSpec.getSalt();
+            String ps = digestAlgorithmParams.toString();
+            kdfHmac = getKdfHmac(ps);
+            if (!(kdfHmac.equals("HmacSHA512") ||
+                    kdfHmac.equals("HmacSHA256"))) {
+                throw new IllegalArgumentException("unsupported PBMAC1 Hmac");
```

Comment:
Hmm. I don't think we need to limit the algorithms we support on loading. It is more about when storing new keystores since the current PBMAC1 algorithm syntax doesn't allow a mix of different Mac/Prf algs.

Also,`engineLoad` is not defined to throw `IllegalArgumentException`, so if this is still needed, then `IOException` is probably more appropriate.

Reply:
We don't have a `PBMAC1ParameterSpec` to pass key length from `PBMAC1Parameters` to `PKCS12KeyStore`. By the time you get to `PKCS12KeyStore`, the only hint for key length is the suffix on the Hmac. I think we need this check, otherwise we will end up with a not very useful "failed PKCS12 integrity checking" message.

I'll change `IllegalArgumentException` to `IOException`.